### PR TITLE
all: add support for thunderbolt networking

### PR DESCRIPTION
**Pull request by gicmo**
First attempt to implement thunderbolt networking. Comments welcome.

Load the thunderbolt-net module if we see a host-to-host connection
and configure the resulting ethernet connection automatically to be
a link-local only one.

### CODE REVIEW

**Comment by vhumpa** [comment]
NM CI: Please confirm test execution over this PR. Comment .*ok\W+to\W+test.* to confirm and/or .*add\W+to\W+whitelist.* to add the PR author in whitelist for future.

File: data/90-nm-thunderbolt.rules [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+ACTION!="add", GOTO="nm_thunderbolt_end"
+
+# Load he thunderbolt-net driver if we a device of type thunderbolt_xdomain
+# is added.
+SUBSYSTEM=="thunderbolt", ENV{DEVTYPE}=="thunderbolt_xdomain", RUN{builtin}+="kmod load thunderbolt-net"
+
+# For all thunderbolt network devices, we want to enable link-local configuration
+SUBSYSTEM=="net", ENV{ID_NET_DRIVER}=="thunderbolt-net", ENV{NM_AUTO_DEFAULT_LINK_ONLY}="1"
+
+LABEL="nm_thunderbolt_end"
```
Comment (thom311) [inline data/90-nm-thunderbolt.rules]: the file must also be marked as `EXTRA_DIST` in `Makefile.am` and marked for installation. Similar, in `data/meson.build`.

File: src/devices/nm-device-ethernet.c [hunk 2]
```diff
@@ -1481,6 +1484,26 @@ new_default_connection (NMDevice *self)
 	g_object_set (setting, NM_SETTING_WIRED_MAC_ADDRESS, perm_hw_addr, NULL);
 	nm_connection_add_setting (connection, setting);
 
+	/* Check if we should create a Link-Local only connection */
+	dev = nm_platform_link_get_udev_device (nm_device_get_platform (NM_DEVICE (self)), nm_device_get_ifindex (self));
```
Comment (thom311) [inline src/devices/nm-device-ethernet.c]: I think this should be `nm_device_get_ip_ifindex (self)`.

File: src/devices/nm-device-ethernet.c [hunk 3]
```diff
@@ -1481,6 +1484,26 @@ new_default_connection (NMDevice *self)
...
+	if (nm_udev_utils_property_as_boolean (uprop)) {
+		setting = nm_setting_ip4_config_new ();
+		g_object_set (setting,
+					  NM_SETTING_IP_CONFIG_METHOD, NM_SETTING_IP4_CONFIG_METHOD_LINK_LOCAL,
+					  NULL);
+		nm_connection_add_setting (connection, setting);
+
+		setting = nm_setting_ip6_config_new ();
+		g_object_set (setting,
+					  NM_SETTING_IP_CONFIG_METHOD,  NM_SETTING_IP6_CONFIG_METHOD_LINK_LOCAL,
```
Comment (thom311) [inline src/devices/nm-device-ethernet.c]: indentation should be with spaces.

**Comment by thom311** [comment]
ah, maybe `NM_AUTO_DEFAULT_LINK_ONLY` should be called `NM_AUTO_DEFAULT_LINK_LOCAL_ONLY`.

**Comment by bengal** [comment]
> ah, maybe NM_AUTO_DEFAULT_LINK_ONLY should be called NM_AUTO_DEFAULT_LINK_LOCAL_ONLY.

+1

Looks good to me, with changes suggested by Thomas.

**Comment by gicmo** [comment]
I hope I got the whitespaces correct. I tested it locally with F27 <-> Windows and successfully got a link-local address that was ping-able from windows.

**Pull request merged**